master/internal/task/all_gather.go

```diff
@@ -0,0 +1,169 @@
+package task
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	"github.com/determined-ai/determined/master/pkg/actor/actors"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+)
+
+// DefaultAllGatherTimeout is the default timeout for all gather.
+var DefaultAllGatherTimeout = 10 * time.Minute
+
+type (
+	// AllGather performs an all gather for an allocation.
+	AllGather struct {
+		id       uuid.UUID
+		watchers map[uuid.UUID]chan<- AllGatherInfoOrError
+		data     []*structpb.Struct
+		numPeers *int
+
+		readyPassed  bool
+		lastPeerJoin time.Time
+	}
+
+	// AllGatherWatcher contains a channel which can be polled for all gather completion.
+	AllGatherWatcher struct {
+		C <-chan AllGatherInfoOrError
```

Comment:
since this actor is returning channels anyway... maybe just make this entire chunk of code without actors, in "normal go"?